### PR TITLE
Partial Refunds Backend Validations

### PR DESCRIPTION
**Pull request by priyatham-bhikkumalla**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java [hunk 1]
```diff
@@ -234,4 +236,30 @@ private String getFeeIds(List<PaymentFee> paymentFees) {
             .collect(Collectors.joining(","));
     }
 
+    private void validateRefund(PaymentRefundRequest paymentRefundRequest, List<PaymentFee> paymentFeeList) {
+
+        if(paymentRefundRequest.getRefundAmount().compareTo(BigDecimal.valueOf(0))==0)
```
[the remark this answers is not in the document]
Comment (priyatham-bhikkumalla) [reply api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java:241]: In older versions of java we used to do that, now it is already taken care by javax validation constraints. In PaymentRefundRequest @NotNull annotation on top of refundAmount will do.

File: api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java [hunk 2]
```diff
@@ -234,4 +236,30 @@ private String getFeeIds(List<PaymentFee> paymentFees) {
...
+                        throw new InvalidPartialRefundRequestException("You need to enter a valid number");
+
+                    if(paymentRefundRequest.getRefundAmount().compareTo(feeDto.getApportionAmount())>0)
+                        throw new InvalidPartialRefundRequestException("The amount you want to refund is more than the amount paid");
+
+                    if(feeDto.getVolume()>paymentFee.getVolume())
+                        throw new InvalidPartialRefundRequestException("The quantity you want to refund is more than the available quantity");
+
+                    if(paymentRefundRequest.getRefundAmount().compareTo(BigDecimal.valueOf((long) feeDto.getFeeAmount().intValue() *feeDto.getVolume()))!=0
+                        && feeDto.getVolume()>1)
```
[the remark this answers is not in the document]
Comment (priyatham-bhikkumalla) [reply api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java:259]: In older versions of java we used to do that, now it is already taken care by javax validations constraints. In PaymentRefundRequest class @NotNull annotation on top of refundAmount will do.

[the remark this answers is not in the document]
Comment (priyatham-bhikkumalla) [reply api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java:259]: All the error messages are already provided by BA in the ticket and the developer's are guided to use the same PAY-5381

File: api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java [hunk 3]
```diff
@@ -234,4 +236,30 @@ private String getFeeIds(List<PaymentFee> paymentFees) {
...
+            for (FeeDto feeDto : paymentRefundRequest.getFees()) {
+
+                if (feeDto.getId().intValue() == paymentFee.getId().intValue()){
+
+                    if(feeDto.getVolume()==0)
+                        throw new InvalidPartialRefundRequestException("You need to enter a valid number");
+
+                    if(paymentRefundRequest.getRefundAmount().compareTo(feeDto.getApportionAmount())>0)
+                        throw new InvalidPartialRefundRequestException("The amount you want to refund is more than the amount paid");
+
```
[the remark this answers is not in the document]
Comment (priyatham-bhikkumalla) [reply api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java:254]: All the error messages are already provided by BA in the ticket and the developer's are guided to use the same PAY-5381

File: api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java [hunk 4]
```diff
@@ -234,4 +236,30 @@ private String getFeeIds(List<PaymentFee> paymentFees) {
...
+        for(PaymentFee paymentFee : paymentFeeList){
+            for (FeeDto feeDto : paymentRefundRequest.getFees()) {
+
+                if (feeDto.getId().intValue() == paymentFee.getId().intValue()){
+
+                    if(feeDto.getVolume()==0)
+                        throw new InvalidPartialRefundRequestException("You need to enter a valid number");
+
+                    if(paymentRefundRequest.getRefundAmount().compareTo(feeDto.getApportionAmount())>0)
```
[the remark this answers is not in the document]
Comment (priyatham-bhikkumalla) [reply api/src/main/java/uk/gov/hmcts/payment/api/service/PaymentRefundsServiceImpl.java:252]: All the error messages are already provided by BA in the ticket and the developer's are guided to use the same PAY-5381

**Review by sathishkannan-hmcts — APPROVED** [review]

**Review by sayali-salunkhe — COMMENTED** [review]
Please merge this PR in **refunds-feature-base-v2** branch once approved as waystopay_feature_base is getting deprecated

**Pull request merged**